src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -24,120 +28,144 @@ public static int CacheSize
                 }
             }
         }
 
         /// <summary>
-        /// Find code cache based on options+pattern
+        ///  Find cache based on options+pattern+culture and optionally add new cache if not found
         /// </summary>
-        private static CachedCodeEntry LookupCachedAndUpdate(CachedCodeEntryKey key)
+        private CachedCodeEntry GetCachedCode(CachedCodeEntryKey key, bool isToAdd)
         {
+            if (s_cacheSize == 0)
+                return null;
             lock (s_livecode)
             {
-                for (LinkedListNode<CachedCodeEntry> current = s_livecode.First; current != null; current = current.Next)
+                // first look for it in the cache and move it to the head
+                var entry = LookupCachedAndPromote(key);
+                // it wasn't in the cache, so we'll add a new one.  Shortcut out for the case where cacheSize is zero.
+                if (entry == null && isToAdd && s_cacheSize != 0)  // check cache size again in case it changed
                 {
-                    if (current.Value._key == key)
+                    entry = new CachedCodeEntry(key, capnames, capslist, _code, caps, capsize, _runnerref, _replref);
+                    s_livecode.Add(key, entry);
+                    // put in linked list:
+                    if (s_livecode_first != null)
+                    {
+                        s_livecode_first._next = entry;
+                        entry._previous = s_livecode_first;
+                    }
+                    s_livecode_first = entry;
+                    if (s_livecode_last == null)
+                    {
+                        s_livecode_last = entry;
+                    }
+                    else
                     {
-                        // If we find an entry in the cache, move it to the head at the same time.
-                        s_livecode.Remove(current);
-                        s_livecode.AddFirst(current);
-                        return current.Value;
+                        if (s_livecode.Count > s_cacheSize)
+                        {
+                            var last = s_livecode_last;
+                            s_livecode.Remove(last._key);
+
+                            last._next._previous = null;
+                            s_livecode_last = last._next;
+                            last._next = null;
+                        }
                     }
                 }
+                return entry;
             }
-
-            return null;
         }
 
-        /// <summary>
-        /// Add current code to the cache
-        /// </summary>
-        private CachedCodeEntry CacheCode(CachedCodeEntryKey key)
+        private static CachedCodeEntry LookupCachedAndPromote(CachedCodeEntryKey key)
         {
-            CachedCodeEntry newcached = null;
-
-            lock (s_livecode)
+            if (s_livecode_first?._key == key) // most used regex should be at the top already
```

Comment:
It still had to take the lock to read this first element. Is that necessary? Reads of pointers are atomic. My thought was you could check outside the lock.

Reply:
I think it is not correct as some other thread can replace `s_livecode_first` after one reads it. And `Interlocked` constructs can't be used as we need to check key there. Or I miss something?

Reply:
I figured that on CachedCodeEntry, _key is never modified. So it is fine to get a pointer to the first one, check the key at our leisure and hand it back? ie
```c#

        /// <summary>
        ///  Find cache based on options+pattern+culture and optionally add new cache if not found
        /// </summary>
        private CachedCodeEntry GetCachedCode(CachedCodeEntryKey key, bool isToAdd)
        {
            if (s_cacheSize == 0)
                return null;
            CachedCodeEntry first = s_livecode_first;
            if (first?._key == key)
                 return first;
            lock (s_livecode)
            {
...
```

Does that seem right?

Reply:
If we mark _key as readonly.

Reply:
@ViktorHofer does this make sense?

Reply:
With this 
```
        private CachedCodeEntry GetCachedCode(CachedCodeEntryKey key, bool isToAdd)
        {
            // to avoid lock:
            if (s_cacheSize == 0)
                return null;
            CachedCodeEntry first = s_livecode_first;
            if (first != null)
                if (first.Key == key)
                    return first;
                else if (!isToAdd && s_livecode.Count == 1) // Not the only one
                    return null;
            else if (!isToAdd) // first == null, empty cache
                return null;
```

the time increased for 800 and 8000 as well as space. Rechecking it.

Reply:
First == null && cachesize != 0 is a rare case, it can be handled inside the lock. Similarly count == 1 is a rare case and need not have special treatment either.

Also since cachesize ==0 is rare, it can be checked second, although I doubt that matters. So can you test with this below?

```
        /// <summary>
        ///  Find cache based on options+pattern+culture and optionally add new cache if not found
        /// </summary>
        private CachedCodeEntry GetCachedCode(CachedCodeEntryKey key, bool isToAdd)
        {
            CachedCodeEntry first = s_livecode_first;
            if (first?._key == key)
                 return first;
            if (s_cacheSize == 0)
                return null;
            lock (s_livecode)
            {
...
```

Reply:
Incidentally, it seems like it must be really rare to use 40_000 distinct regexes in an app, and 400_000 even more so. If there are that many, at a guess they would be generated, and not reused - so a cache is just overhead.

I'm gathering some data from internal use that I should be able to share soon, but my guess is it would be more realistic to test with 100-1000 regexes at most. That's a pure guess and guesses can be wrong...